clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -65,17 +66,91 @@
  */
 public class TransactionManager {
     private static final int NO_INFLIGHT_REQUEST_CORRELATION_ID = -1;
+    private static final int NO_LAST_ACKED_SEQUENCE_NUMBER = -1;
 
     private final Logger log;
     private final String transactionalId;
     private final int transactionTimeoutMs;
 
-    // The base sequence of the next batch bound for a given partition.
-    private final Map<TopicPartition, Integer> nextSequence;
+    private static class TopicPartitionBookkeeper {
 
-    // The sequence of the last record of the last ack'd batch from the given partition. When there are no
-    // in flight requests for a partition, the lastAckedSequence(topicPartition) == nextSequence(topicPartition) - 1.
-    private final Map<TopicPartition, Integer> lastAckedSequence;
+        private final Map<TopicPartition, TopicPartitionEntry> topicPartitionBookkeeping = new HashMap<>();
+
+        void initSequenceNumber(TopicPartition topicPartition) {
+            get(topicPartition).nextSequenceNumber = 0;
+        }
+
+        boolean hasSequenceNumber(TopicPartition topicPartition) {
+            TopicPartitionEntry entry = topicPartitionBookkeeping.get(topicPartition);
+            return entry != null && entry.nextSequenceNumber != null;
+        }
+
+        boolean hasPendingOffsetCommits() {
+            return topicPartitionBookkeeping.values().stream().anyMatch(tpe -> tpe.pendingTxnOffsetCommit != null);
+        }
+
+        void clearPendingOffsetCommits() {
+            topicPartitionBookkeeping.values().forEach(tpe -> tpe.pendingTxnOffsetCommit = null);
+        }
+
+        void clearPendingOffsetCommit(TopicPartition topic) {
+            TopicPartitionEntry tpe = topicPartitionBookkeeping.get(topic);
+            if (tpe != null) {
+                tpe.pendingTxnOffsetCommit = null;
+            }
+        }
+
+        Map<TopicPartition, CommittedOffset> pendingOffsetCommits() {
+            return topicPartitionBookkeeping.entrySet()
+                    .stream()
+                    .filter(e -> e.getValue() != null && e.getValue().pendingTxnOffsetCommit != null)
```

Comment:
Why do we need `e -> e.getValue() != null`?

Reply:
It seems to have no effect, I removed them. Probably just had the Pavlovian reflex for null checks on objects. :)